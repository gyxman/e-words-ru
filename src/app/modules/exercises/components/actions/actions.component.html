<button
    automation-id="submit-button"
    mat-raised-button
    color="primary"
    class="button"
    (click)="onSubmit()"
    [disabled]="disabledSubmit"
>
    Проверить
</button>
