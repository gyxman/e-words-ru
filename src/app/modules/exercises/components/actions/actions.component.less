:host {
    display: block;
    max-width: 200px;
}

.button {
    width: 100%;
}
