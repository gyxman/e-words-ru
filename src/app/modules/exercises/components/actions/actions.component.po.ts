import {PageObject} from '../../../../testing/page-object';
import {DebugElement} from '@angular/core';

export class ActionsComponentPo<T> extends PageObject<T> {
    get submitButton(): DebugElement {
        return this.getByAutomationId('submit-button');
    }
}
