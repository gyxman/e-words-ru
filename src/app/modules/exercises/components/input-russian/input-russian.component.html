<ng-container *ngIf="word$ | async as word">
    <app-loader automation-id="loader" [showLoader]="showLoader$ | async">
        <div class="form">
            <h3 class="title">Введите перевод для слова:</h3>
            <div class="word">{{ word | mapper: getCurrentWord }}</div>
            <mat-form-field class="form-field">
                <mat-label>Ответ</mat-label>
                <input
                    automation-id="answer-input"
                    matInput
                    placeholder="Введите слово на русском языке"
                    [formControl]="answerControl"
                    (keydown.enter)="onSubmit()"
                />
                <app-form-control-error
                    class="form-error"
                    [control]="answerControl"
                ></app-form-control-error>
            </mat-form-field>
            <app-actions
                automation-id="actions"
                class="actions"
                [disabledSubmit]="!answerControl.value"
                (submit)="onSubmit()"
            ></app-actions>
        </div>
    </app-loader>
</ng-container>
