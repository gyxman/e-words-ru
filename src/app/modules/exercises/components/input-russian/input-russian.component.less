@import '../../../../../styles/index';

:host {
    display: block;
}

.form {
    width: 100%;
    max-width: 400px;
    margin: 0 auto;
    text-align: center;
}

.title {
    .font-p1();
    font-weight: normal;
    margin: 0;
}

.word {
    .font-h1();
    margin: 30px 0 20px;
}

.form-error {
    position: absolute;
    bottom: -20px;
    left: 0;
}

.actions {
    margin: 20px auto 0;
}
