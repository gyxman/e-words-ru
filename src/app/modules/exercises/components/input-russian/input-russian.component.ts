import {
    ChangeDetectionStrategy,
    Component,
    EventEmitter,
    Input,
    Output,
} from '@angular/core';
import {Word} from '../../../words/models/word';
import {Observable} from 'rxjs';
import {FormControl, Validators} from '@angular/forms';

@Component({
    selector: 'app-input-russian',
    templateUrl: './input-russian.component.html',
    styleUrls: ['./input-russian.component.less'],
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class InputRussianComponent {
    @Input() word$: Observable<Word>;
    @Input() showLoader$: Observable<boolean>;

    @Output() answer = new EventEmitter<string>();

    answerControl = new FormControl(null, [Validators.required]);

    getCurrentWord({englishWord}: Word): string {
        return englishWord;
    }

    onSubmit() {
        this.answer.emit(this.answerControl.value.trim());
        this.answerControl.patchValue(null);
    }
}
