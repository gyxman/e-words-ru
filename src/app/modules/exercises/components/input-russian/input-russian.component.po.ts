import {PageObject} from '../../../../testing/page-object';
import {DebugElement} from '@angular/core';
import {MockedComponent} from 'ng-mocks';
import {LoaderComponent} from '../../../utils/components/loader/loader.component';
import {ActionsComponent} from '../actions/actions.component';

export class InputRussianComponentPo<T> extends PageObject<T> {
    get loaderComponent(): MockedComponent<LoaderComponent> {
        return this.getByAutomationId('loader').componentInstance;
    }

    get answerInput(): HTMLInputElement {
        return this.getByAutomationId('answer-input').nativeElement;
    }

    get actions(): DebugElement {
        return this.getByAutomationId('actions');
    }

    get actionsComponent(): MockedComponent<ActionsComponent> {
        return this.getByAutomationId('actions').componentInstance;
    }
}
