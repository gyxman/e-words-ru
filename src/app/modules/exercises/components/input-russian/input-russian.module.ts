import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {InputRussianComponent} from './input-russian.component';
import {MatInputModule} from '@angular/material/input';
import {ReactiveFormsModule} from '@angular/forms';
import {FormControlErrorModule} from '../../../auth/components/form-control-error/form-control-error.module';
import {MapperModule} from '../../../utils/pipes/mapper/mapper.module';
import {ActionsModule} from '../actions/actions.module';
import {LoaderModule} from '../../../utils/components/loader/loader.module';

@NgModule({
    imports: [
        CommonModule,
        MapperModule,
        MatInputModule,
        ReactiveFormsModule,
        FormControlErrorModule,
        ActionsModule,
        LoaderModule,
    ],
    declarations: [InputRussianComponent],
    exports: [InputRussianComponent],
})
export class InputRussianModule {}
