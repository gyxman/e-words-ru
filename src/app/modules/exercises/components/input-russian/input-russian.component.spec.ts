import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {InputRussianComponent} from './input-russian.component';
import {Component} from '@angular/core';
import {BehaviorSubject, ReplaySubject} from 'rxjs';
import {Word} from '../../../words/models/word';
import {InputRussianComponentPo} from './input-russian.component.po';
import {LoaderModule} from '../../../utils/components/loader/loader.module';
import {MapperModule} from '../../../utils/pipes/mapper/mapper.module';
import {MatInputModule} from '@angular/material/input';
import {ReactiveFormsModule} from '@angular/forms';
import {FormControlErrorModule} from '../../../auth/components/form-control-error/form-control-error.module';
import {ActionsModule} from '../actions/actions.module';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';

@Component({
    template:
        '<app-input-russian [word$]="word$" [showLoader$]="showLoader$" (answer)="onAnswer($event)"></app-input-russian>',
})
class TestComponent {
    word$: BehaviorSubject<Word> = new BehaviorSubject({
        id: 'wordId',
        englishWord: 'testWord',
    } as Word);
    showLoader$: ReplaySubject<boolean> = new ReplaySubject(1);

    answer: string;

    onAnswer(answer: string) {
        this.answer = answer;
    }
}

describe('InputRussianComponent - упражнение, где пользователь вводит перевод слова на русском', () => {
    let testComponent: TestComponent;
    let fixture: ComponentFixture<TestComponent>;
    let pageObject: InputRussianComponentPo<TestComponent>;

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            declarations: [InputRussianComponent, TestComponent],
            imports: [
                LoaderModule,
                MapperModule,
                MatInputModule,
                ReactiveFormsModule,
                FormControlErrorModule,
                ActionsModule,
                NoopAnimationsModule,
            ],
        }).compileComponents();
    }));

    beforeEach(() => {
        fixture = TestBed.createComponent(TestComponent);
        testComponent = fixture.componentInstance;
        pageObject = new InputRussianComponentPo(fixture);
    });

    it('Если компонент загрузился, все элементы отрисовались', () => {
        // assert
        expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('Если пришла информация об отображении лоадера, то отображаем лоадер', () => {
        // arrange
        testComponent.showLoader$.next(true);

        // act
        fixture.detectChanges();

        // assert
        expect(pageObject.loaderComponent.showLoader).toBe(true);
    });

    it('Если информация об отображении лоадера не пришла, то не отображаем лоадер', () => {
        // arrange
        testComponent.showLoader$.next(false);

        // act
        fixture.detectChanges();

        // assert
        expect(pageObject.loaderComponent.showLoader).toBe(false);
    });

    it('Если компонент загрузился, то в поле ответ пусто', () => {
        // act
        fixture.detectChanges();

        // assert
        expect(pageObject.answerInput.value).toBe('');
    });

    it('Если компонент загрузился, то в компонент actions передаем информацию о блокировке кнопки отправки', () => {
        // act
        fixture.detectChanges();

        // assert
        expect(pageObject.actionsComponent.disabledSubmit).toBe(true);
    });

    it('Если пользователь вводит ответ и нажимает на кнопку отправки ответа, то вызываем метод отправки ответа', () => {
        // arrange
        jest.spyOn(testComponent, 'onAnswer');
        fixture.detectChanges();

        // act
        pageObject.answerInput.value = 'ответ';
        pageObject.answerInput.dispatchEvent(new Event('input'));
        fixture.detectChanges();
        pageObject.actions.triggerEventHandler('submit', null);

        // assert
        expect(testComponent.onAnswer).toHaveBeenCalledWith('ответ');
    });
});
