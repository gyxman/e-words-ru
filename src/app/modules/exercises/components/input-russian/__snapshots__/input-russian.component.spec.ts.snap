// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRussianComponent - упражнение, где пользователь вводит перевод слова на русском Если компонент загрузился, все элементы отрисовались 1`] = `
<div
  id="root0"
>
  <app-input-russian>
    
  </app-input-russian>
</div>
`;
